central/graphql/resolvers/image_vulnerabilities.go

```diff
@@ -101,3 +199,323 @@ func withImageCveTypeFiltering(q string) string {
+	osLevelComponents, err := resolver.root.ImageComponentDataStore.Count(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	if osLevelComponents == 0 {
+		return &activeStateResolver{root: resolver.root, state: Undetermined}, nil
+	}
+
+	qb := search.NewQueryBuilder().AddExactMatches(search.DeploymentID, deploymentID)
+	imageID := getImageIDFromQuery(scopeQuery)
+	if imageID != "" {
+		qb.AddExactMatches(search.ImageSHA, imageID)
+	}
+	query = search.ConjunctionQuery(resolver.getImageCVEQuery(), qb.ProtoQuery())
+
+	results, err := resolver.root.ActiveComponent.Search(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	ids := search.ResultsToIDs(results)
+	state := Inactive
+	if len(ids) != 0 {
+		state = Active
+	}
+	return &activeStateResolver{root: resolver.root, state: state, activeComponentIDs: ids, imageScope: imageID}, nil
+}
+
+func (resolver *imageCVEResolver) EffectiveVulnerabilityRequest(ctx context.Context) (*VulnerabilityRequestResolver, error) {
+	var imageID string
+	scope, hasScope := scoped.GetScopeAtLevel(resolver.ctx, v1.SearchCategory_IMAGES)
+	if hasScope {
+		imageID = scope.ID
+	}
+
+	if imageID == "" {
+		return nil, errors.Errorf("image scope must be provided for determining effective vulnerability request for cve %s", resolver.data.GetId())
+	}
+	imageLoader, err := loaders.GetImageLoader(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "getting image loader")
+	}
+	img, err := imageLoader.FromID(ctx, imageID)
+	if err != nil {
+		log.Error(errors.Wrapf(err, "fetching image with id %s", imageID))
+		return nil, nil
+	}
+
+	req, err := resolver.root.vulnReqQueryMgr.EffectiveVulnReq(ctx, resolver.data.GetId(),
+		common.VulnReqScope{
+			Registry: img.GetName().GetRegistry(),
+			Remote:   img.GetName().GetRemote(),
+			Tag:      img.GetName().GetTag(),
+		})
+	if err != nil {
+		return nil, err
+	}
+	return resolver.root.wrapVulnerabilityRequest(req, nil)
+}
+
+func (resolver *imageCVEResolver) DeploymentCount(ctx context.Context, args RawQuery) (int32, error) {
+	return resolver.root.DeploymentCount(resolver.withImageVulnerabilityScope(ctx), args)
+}
+
+func (resolver *imageCVEResolver) Deployments(ctx context.Context, args PaginatedQuery) ([]*deploymentResolver, error) {
+	return resolver.root.Deployments(resolver.withImageVulnerabilityScope(ctx), args)
+}
+
+func (resolver *imageCVEResolver) DiscoveredAtImage(ctx context.Context, args RawQuery) (*graphql.Time, error) {
+	var imageID string
+	scope, hasScope := scoped.GetScopeAtLevel(ctx, v1.SearchCategory_IMAGES)
+	if hasScope {
+		imageID = scope.ID
+	} else {
+		var err error
+		imageID, err = getImageIDFromIfImageShaQuery(ctx, resolver.root, args)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not determine vulnerability discovered time in image")
+		}
+	}
+
+	if imageID == "" {
+		return nil, nil
+	}
+
+	edgeID := edges.EdgeID{ParentID: imageID, ChildID: resolver.data.GetId()}.ToString()
```

Comment:
- This will not work on postgres
- We should avoid constructing IDs at service layer. Instead, use query.

Reply:
`postgres.IDFromPKs`